smbios: add enums/bitflags types for some SMBIOS fields

hawkw opened the pull request:
Currently, most of the SMBIOS fields that we populate are magic numbers, with comments describing what that magic number indicates. For SMBIOS enum fields, this isn't too bad, but for bitfields like the various "characteristics" fields, it's a bit of a shame, since the hex constant doesn't always make it obvious which bit corresponds to which characteristic.

To make this a bit nicer, this commit adds enums for most of the SMBIOS enum fields we currently populate, and bitflags for most of the bitflags fields. This way, we can refer to them by name rather than as magic numbers. Also, I've made the `Default` implementation for the various enum fields return the "unknown" (0x2) variant. Eventually, we may want to default-initialize more SMBIOS tables --- but, we should probably hold off on that until *all* the enum fields in the tables we populate are Rust types with `Default` impls, so that we don't have to manually initialize some of them to 0x2.

pfmooney commented on bin/propolis-server/src/lib/initializer.rs:
```diff
@@ -864,9 +865,10 @@ impl<'a> MachineInitializer<'a> {
                 .unwrap(),
             bios_rom_size: ((rom_size / (64 * 1024)) - 1) as u8,
             // Characteristics-not-supported
```
style nit: Since you're turning these into nice named values, the accompanying comments can probably be tossed out now (here and below)

pfmooney commented on bin/propolis-server/src/lib/initializer.rs:
```diff
@@ -864,9 +865,10 @@ impl<'a> MachineInitializer<'a> {
-            bios_characteristics: 0x8,
-            // ACPI + UEFI + IsVM
-            bios_ext_characteristics: 0x1801,
+            bios_characteristics: type0::BiosCharacteristics::UNKNOWN,
```
Here and in -server, this should be not-supported, rather than unknown.

pfmooney commented on lib/propolis/src/firmware/smbios/table.rs:
```diff
@@ -138,19 +324,121 @@ impl Table for Type4 {
+        #[default]
+        Unknown = 0x02,
+        /// Central Processor
+        Cpu = 0x03,
```
nit: Why Cpu instead of Central, to stay consistent with the other entries?

pfmooney commented on bin/propolis-standalone/src/main.rs:
```diff
@@ -876,31 +879,32 @@ fn generate_smbios(params: SmbiosParams) -> anyhow::Result<smbios::TableBytes> {
+        array_use: type16::ArrayUse::System,
         // unknown
-        error_correction: 0x2,
+        error_correction: type16::ErrorCorrection::Unknown,
```
Omitted for standalone, but not here?

hawkw replied:
ah, the default impl populates this with `Unknown` now, and I went back and
forth on whether I wanted to use the default impl for `Unknown` fields when we
don't default to unknown for stuff that isn't an enum/bitflags yet. I'll go back
and add it in standalone.

pfmooney reviewed (commented):
For testing notes, it'd be nice to see that `dmidecode` (or `smbios` on illumos) output remained unchanged before and after.


Thanks for polishing these bits up

hawkw commented:
@pfmooney: I've addressed your review feedback (thanks for the quick review & suggestions!) and I've added `serde::{Serialize, Deserialize}` impls for all the new types that should serialize/deserialize them as raw numbers and validate when deserializing --- let me know what you think!

I'd also like add a PHD test that runs `dmidecode` in a guest and checks its output, but I'm not sure if it's installed out of the box in our test Alpine image...

pfmooney commented on lib/propolis/src/firmware/smbios/table.rs:
```diff
@@ -184,6 +711,213 @@ impl Table for Type16 {
+        /// See Table 74 in section 7.17.3 of [the SMBIOS Reference
+        /// Specification][DSP0136] for details.
+        ///
+        /// [DSP0136]:
+        ///     https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.7.0.pdf
```
Could probably put this reference at the top somewhere, rather than repeating it at each bitfield.

hawkw commented on lib/propolis/src/firmware/smbios/table.rs:
```diff
@@ -46,14 +57,284 @@ impl Table for Type0 {
+        }
+
+        #[automatically_derived]
+        impl std::convert::TryFrom<$repr> for $Enum {
+            type Error = &'static str;
+            fn try_from(value: $repr) -> Result<Self, Self::Error> {
+                match value {
+                    $val1 => Ok($Enum::$Variant1),
+                    $(
+                        $val => Ok($Enum::$Variant),
+                    )+
+                    _ => Err(concat!(
+                        "invalid ", stringify!($enum), " value, expected one of: ",
+                        stringify!($Variant1)," (", stringify!($val1), ")",
+                        $(", ", stringify!($Variant), " (", stringify!($val), ")" ),*
+                    )),
+                }
+            }
+        }
```
Ah, I didn't realize we already had a dep on `strum`, that'll make this much neater!

pfmooney replied:
A convenient crate, for sure

pfmooney commented on lib/propolis/src/firmware/smbios/table.rs:
```diff
@@ -46,14 +68,303 @@ impl Table for Type0 {
+            fn $name() {
+                for variant in <$Enum>::VARIANTS {
+                    let serialized =
+                        dbg!(serde_json::to_string(&(*dbg!(variant) as $repr)))
+                            .unwrap();
+                    let deserialized =
+                        dbg!(serde_json::from_str::<'_, $Enum>(&serialized))
+                            .unwrap();
+                    assert_eq!(*variant, deserialized);
+                }
+
+                for invalid in $invalid {
+                    let serialized = dbg!(serde_json::to_string(&invalid)).unwrap();
+
+                    dbg!(serde_json::from_str::<'_, $Enum>(&serialized))
+                        .unwrap_err();
+                }
+            }
+        )+
+    }
+}
+
+#[cfg(test)]
+macro_rules! enum_serde_roundtrip_tests {
+    ($(fn $name:ident($Enum:ty) {})+) => {
+        $(
+            #[test]
+            fn $name() {
+                for variant in <$Enum>::VARIANTS {
+                    let serialized =
+                        dbg!(serde_json::to_string(dbg!(variant))).unwrap();
+                    let deserialized =
+                        dbg!(serde_json::from_str::<'_, $Enum>(&serialized))
+                            .unwrap();
+                    assert_eq!(*variant, deserialized);
+                }
+            }
+        )+
+    }
+}
```
Nice to cut down on all that boilerplate with these.

pfmooney commented on lib/propolis/src/firmware/smbios/table.rs:
```diff
@@ -46,14 +68,303 @@ impl Table for Type0 {
+                        static ref ERR_MSG: String = {
+                            use std::fmt::Write;
+                            let mut err = "expected one of: ".to_string();
+                            let mut variants = <$Enum>::VARIANTS.iter().copied().peekable();
+                            let mut first = true;
+                            while let Some(variant) = variants.next() {
+                                let has_remaining = variants.peek().is_some();
+                                let or = if !has_remaining && !first {
+                                    " or "
+                                } else {
+                                    ""
+                                };
+
+                                let comma = if has_remaining {
+                                    ", "
+                                } else {
+                                    ""
+                                };
+                                write!(err, "{or}{variant:?} ({:#04x}){comma}", variant as $repr).unwrap();
+                                first = false;
+                            }
```
Might be simpler to impl this in terms of `slice::join`, switching on `VARIANTS.len()` if you want to preserve the fancy behavior for the 1 and 2 item cases.

Up to you.

pfmooney reviewed (approved):
Looks good.  I assume you've done the `dmidecode` diff testing to ensure that nothing has changed on the guest-visible output?

hawkw commented:
Looks like we're good: 
```console
$ colordiff dmidecode-master.txt dmidecode-branch.txt
colordiff 1.0.21 (http://www.colordiff.org/)
(C)2002-2022 Dave Ewart, davee@sungate.co.uk

```
